akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -0,0 +1,289 @@
+  def apply(system: ActorSystem, serverSettings: ServerSettings): BidiFlow[HttpResponse, HttpResponse, HttpRequest, HttpRequest, ServerTerminator] = {
+
+    val stage = new GracefulTerminatorStage(serverSettings)
+
+    val terminator = new ServerTerminator {
+      override def terminate(deadline: FiniteDuration)(implicit ec: ExecutionContext): Future[HttpTerminated] = {
+        system.log.warning("Called ServerTerminator.terminate() however the stage was not installed! Check your settings.")
+
+        Future.failed(new Exception("Termination support was not enabled on this server"))
+      }
+    }
+
+    BidiFlow.identity[HttpResponse, HttpRequest].mapMaterializedValue(_ ⇒ terminator)
+  }
+}
+
+/**
+ * Implements "graceful" termination request being handled on this connection.
+ *
+ * Termination works as follows:
+ *
+ * 1) unbind:
+ * - the server is unbound; no new connections will be accepted
+ *
+ * 2) handle in-flight request:
+ * - if no request is "in-flight" (being handled by user code)
+ *   - the connection is terminated
+ * - if a request is "in-flight" (being handled by user code), it is given `deadline` time to complete,
+ *   - if user code emits a response within the timeout they do we deliver this response
+ *     - however if it is a streaming response, we also mandate that it shall complete within `deadline` (we will terminate the connection then)
+ *   - if user code does not reply with a response within the deadline, we produce a special terminationDeadlineExceeded HTTP response (e.g. 503 Service Unavailable)
+ *     - the HTTP response emitted by this stage in this case is configurable in [[ServerSettings]]
+ *
+ * 3) close still existing connections
+ * - rather than keeping connections around until idle-timeouts hit (or serving still incoming requests,
+ *   which would risk such incoming requests leading to the server never being "really done processing"),
+ *   all still remaining connections will be forcefully terminated. The termination process is then considered completed.
+ *
+ * Stage shape diagram:
+ *
+ * {{{
+ *                      +---+
+ * fromNet Request   -> | G | -> toUser Request
+ *                      | T |
+ * toNet   Response  <- | S | <- fromUser Response
+ *                      +---+
+ * }}}
+ */
+final class GracefulTerminatorStage(settings: ServerSettings) extends GraphStageWithMaterializedValue[BidiShape[HttpResponse, HttpResponse, HttpRequest, HttpRequest], ServerTerminator] {
+
+  val fromNet: Inlet[HttpRequest] = Inlet("netIn")
+  val toUser: Outlet[HttpRequest] = Outlet("userOut")
+
+  val fromUser: Inlet[HttpResponse] = Inlet("userIn")
+  val toNet: Outlet[HttpResponse] = Outlet("netOut")
+  override def shape = BidiShape(fromUser, toNet, fromNet, toUser)
+
+  final val TerminationDeadlineTimerKey = "TerminationDeadlineTimerKey"
+
+  final class ConnectionTerminator(triggerTermination: Promise[FiniteDuration ⇒ Future[HttpTerminated]]) extends ServerTerminator {
```

Comment:
Pretty intricate logic here.. can be at least a little simpler by taking a `triggerTermination: Future[FiniteDuration ⇒ Future[HttpTerminated]]` here